app/api/search/search.js

```diff
@@ -284,57 +293,154 @@ const _getAggregationDictionary = async (aggregation, language, property, dictio
     return [dictionary, indexedDictionaryValues(dictionary)];
   }
 
-  const dictionary = dictionaries.find(d => d._id.toString() === property.content.toString());
-  return [dictionary, indexedDictionaryValues(dictionary)];
+  const propContent = property.content.toString();
+  if (!dictionaryCache[propContent]) {
+    const dictionary = dictionariesById[propContent];
+    const dictionaryValues = indexedDictionaryValues(dictionary);
+    dictionaryCache[propContent] = [dictionary, dictionaryValues];
+  }
+  return dictionaryCache[propContent];
+};
+
+const extractMissingBucket = buckets => {
+  const [missingBuckets, remainingBuckets] = _.partition(buckets, b => b.key === 'missing');
+  const missingBucket = missingBuckets && missingBuckets.length ? missingBuckets[0] : null;
+  return { missingBucket, remainingBuckets };
 };
 
-const _formatDictionaryWithGroupsAggregation = (aggregation, dictionary) => {
-  const buckets = dictionary.values
-    .map(dictionaryValue => {
-      const bucket = aggregation.buckets.find(b => b.key === dictionaryValue.id);
-      if (bucket && dictionaryValue.values) {
+const groupAndLimitBuckets = (buckets, dictionary, _limit) => {
+  const aggregationBucketsByKey = objectIndex(
+    buckets,
+    b => b.key,
+    b => b
+  );
+  const missingBucket = aggregationBucketsByKey.missing;
+  const limit = missingBucket ? _limit - 1 : _limit;
+  const newBuckets = [];
+
+  let dictIndex = 0;
+  while (newBuckets.length < limit && dictIndex < dictionary.values.length) {
+    const dictionaryValue = dictionary.values[dictIndex];
+    const bucket = aggregationBucketsByKey[dictionaryValue.id];
+    if (bucket) {
+      if (dictionaryValue.values) {
         bucket.values = dictionaryValue.values
-          .map(v => aggregation.buckets.find(b => b.key === v.id))
+          .map(v => aggregationBucketsByKey[v.id])
           .filter(b => b);
       }
-      return bucket;
-    })
-    .filter(b => b);
-  const bucketsIncludeMissing = aggregation.buckets.find(b => b.key === 'missing');
-  if (bucketsIncludeMissing) {
-    buckets.push(bucketsIncludeMissing);
+      newBuckets.push(bucket);
+    }
+    dictIndex += 1;
   }
-  return Object.assign(aggregation, { buckets });
+
+  if (missingBucket) newBuckets.push(missingBucket);
+  return newBuckets;
+};
+
+const limitBuckets = (buckets, _limit) => {
```

Comment:
I know we have this kind of uses of the `_` prefix in the codebase, but I would try to move away from it since it's also the convention for unused paramenters.